src/Ansatz/Chain.jl

```diff
@@ -112,11 +112,19 @@ function Chain(::Operator, boundary::Open, arrays::Vector{<:AbstractArray})
     Chain(Quantum(TensorNetwork(_tensors), sitemap), boundary)
 end
 
-rightsite(tn::Chain, site::Site) = rightsite(boundary(tn), tn, site)
-rightsite(::Union{Open, Periodic}, tn::Chain, site::Site) = Site(site.id + 1)
-
 leftsite(tn::Chain, site::Site) = leftsite(boundary(tn), tn, site)
-leftsite(::Union{Open, Periodic}, tn::Chain, site::Site) = Site(site.id - 1)
+function leftsite(::Open, tn::Chain, site::Site)
+    (site.id > length(sites(tn)) || site.id <= 1) && throw(ArgumentError("Invalid site $site"))
+    Site(site.id - 1)
+end
+leftsite(::Periodic, tn::Chain, site::Site) = Site(mod1(site.id - 1, length(sites(tn))))
+
+rightsite(tn::Chain, site::Site) = rightsite(boundary(tn), tn, site)
+function rightsite(::Open, tn::Chain, site::Site)
+    (site.id > length(sites(tn))-1 || site.id < 1) && throw(ArgumentError("Invalid site $site"))
+    Site(site.id + 1)
+end
+rightsite(::Periodic, tn::Chain, site::Site) = Site(mod1(site.id + 1, length(sites(tn))))
```

Comment:
Same as above: check that `site` is in range.

Reply:
Done!